homeassistant/components/freebox/router.py

```diff
@@ -0,0 +1,222 @@
+"""Represent the Freebox router and its devices and sensors."""
+from datetime import datetime, timedelta
+import logging
+from pathlib import Path
+from typing import Dict
+
+from aiofreepybox import Freepybox
+from aiofreepybox.api.system import System
+from aiofreepybox.api.wifi import Wifi
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST, CONF_PORT
+from homeassistant.helpers.device_registry import CONNECTION_NETWORK_MAC
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.event import async_track_time_interval
+from homeassistant.helpers.typing import HomeAssistantType
+from homeassistant.util import slugify
+
+from .const import (
+    API_VERSION,
+    APP_DESC,
+    CONNECTION_SENSORS,
+    DOMAIN,
+    STORAGE_KEY,
+    STORAGE_VERSION,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = timedelta(seconds=30)
+
+
+class FreeboxRouter:
+    """Representation of a Freebox router."""
+
+    def __init__(self, hass: HomeAssistantType, entry: ConfigEntry):
+        """Initialize a Freebox router."""
+        self.hass = hass
+        self._entry = entry
+        self._host = entry.data[CONF_HOST]
+        self._port = entry.data[CONF_PORT]
+
+        self._api: Freepybox = None
+        self._name = None
+        self._mac = None
+        self._sw_v = None
+        self._attrs = {}
+
+        # Devices
+        self._devices: Dict[str, any] = {}
+
+        # Sensors
+        self._temperature_sensors: Dict[str, any] = {}
+        self._connection_sensors: Dict[str, any] = {}
+
+        self.listeners = []
+
+    async def setup(self) -> None:
+        """Set up a Freebox router."""
+        self._api = await get_api(self.hass, self._host)
+
+        await self._api.open(self._host, self._port)
+
+        # System
+        fbx_config = await self._api.system.get_config()
+        self._mac = fbx_config["mac"]
+        self._name = fbx_config["model_info"]["pretty_name"]
+        self._sw_v = fbx_config["firmware_version"]
+
+        # Devices & sensors
+        await self.update_all()
+        async_track_time_interval(self.hass, self.update_all, SCAN_INTERVAL)
+
+    async def update_all(self, now=None) -> None:
```

Comment:
`now` can also be a datetime instance.

Reply:
?

I don't need `now`

Reply:
It's still passed by the event listener.

Reply:
I still don’t get your comment

Reply:
`update_all` is used as a callback for `async_track_time_interval`. That will pass the event time as the `now` parameter when calling the callback.

https://github.com/home-assistant/home-assistant/blob/16dba3fa8519151cff4064fcaccfb194e4bd0ba0/homeassistant/helpers/event.py#L274-L291

To make the typing correct we need to mark `now` as `Optional[datetime]`.

Reply:
I know that `async_track_time_interval ` is passing a `datetime` as the second parameter of its callback.
I didn't get you where talking about the typing!

Perhaps I don't think it's necessary to type an unused parameter, and it will import `Optional` only for that unused parameter.
I am more willing to modify the `interval_listener` to make this second `datetime` parameter optional.
While checking in the code, I've seen a lot of `now=None` or just `event_time`, all not used in the callback function!

This might also work: (`datetime` already imported)
```suggestion
    async def update_all(self, now:datetime =None) -> None:
```
I can do that if you want

Reply:
Please either type the function correctly or remove the typing completely.